test_othermlpackages/test_pytorch.py

```diff
@@ -0,0 +1,97 @@
+# Copyright (c) Microsoft Corporation and Fairlearn contributors.
+# Licensed under the MIT License.
+
+import numpy as np
```

Comment:
When I try running this file, I wind up with an error
```
E   ImportError: DLL load failed while importing statlib: Not enough memory resources are available to process this command.
```
Any idea what causes that? I'm prodding search engines, but not turning up anything useful.

Reply:
Hmm I do not know.
Some questions for debugging:
Are you on Windows/Linux/Mac? 
Are you using the CPU or GPU version?
Do you have a gpu with CUDA installed and nvidia-drivers?

Reply:
I was doing this on Windows, after installing pytorch with `conda install pytorch cpuonly -c pytorch`. I'm going to grab an Ubuntu VM and try there, since the actual build runs on Ubuntu.

Reply:
Oh wait! I put python==3.8 by mistake before pushing. Pytorch currently has a bug with that version. I'll downgrade and test locally.

Reply:
OK. I have also tried on the Ubuntu VM, and the tests error out with:
```
package_test_common.py:62: in run_expgrad_classification
    expgrad.fit(X_train, Y_train, sensitive_features=A_train)
../fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:122: in fit
    h, h_idx = lagrangian.best_h(lambda_vec)
../fairlearn/reductions/_exponentiated_gradient/_lagrangian.py:190: in best_h
    classifier = self._call_oracle(lambda_vec)
../fairlearn/reductions/_exponentiated_gradient/_lagrangian.py:178: in _call_oracle
    estimator.fit(self.X, redY, **{self.sample_weight_name: redW})
test_pytorch.py:41: in fit
    y = y.reshape([-1, 1])
```

Also, do you need to add `conda-forge` to the list of repositories in order to install `skorch`?

Reply:
It is fixed in my lastest version

Reply:
I just got your last push (must have done my checkout minutes before...), and it's looking more promising. Still a bit concerned about the conda env file not working, though.

Reply:
Thanks for the updates. I just tried doing the same as the build will - create a conda env, and then pip install fairlearn. However, that is failing during a pickle step. I'm trying to attach the error output, and the conda and pip environment information.

Reply:
Doing this the long way.... this is std.out from `pytest`:

```python
============================= test session starts ==============================
platform linux -- Python 3.8.0, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/rge21/fairlearn
collected 3 items

test_pytorch.py FFF                                                      [100%]

=================================== FAILURES ===================================
_________________________ test_expgrad_classification __________________________

    def test_expgrad_classification():
        estimator = create_model()
        disparity_moment = DemographicParity()
    
>       ptc.run_expgrad_classification(estimator, disparity_moment)

test_pytorch.py:86: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
package_test_common.py:56: in run_expgrad_classification
    unmitigated = copy.deepcopy(estimator)
../../miniconda3/envs/skorch-2/lib/python3.8/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
../../miniconda3/envs/skorch-2/lib/python3.8/copy.py:271: in _reconstruct
    y.__setstate__(state)
../../miniconda3/envs/skorch-2/lib/python3.8/site-packages/skorch/net.py:1696: in __setstate__
    cuda_attrs = torch.load(f, **load_kwargs)
../../miniconda3/envs/skorch-2/lib/python3.8/site-packages/torch/serialization.py:529: in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

f = <tempfile.SpooledTemporaryFile object at 0x7fb62901eb50>
map_location = device(type='cpu')
pickle_module = <module 'pickle' from '/home/rge21/miniconda3/envs/skorch-2/lib/python3.8/pickle.py'>
pickle_load_args = {'encoding': 'utf-8'}
legacy_load = <function _legacy_load.<locals>.legacy_load at 0x7fb629025b80>
persistent_load = <function _legacy_load.<locals>.persistent_load at 0x7fb629025d30>
f_should_read_directly = True

    def _legacy_load(f, map_location, pickle_module, **pickle_load_args):
        deserialized_objects = {}
    
        restore_location = _get_restore_location(map_location)
    
        def _check_container_source(container_type, source_file, original_source):
            try:
                current_source = ''.join(get_source_lines_and_file(container_type)[0])
            except Exception:  # saving the source is optional, so we can ignore any errors
                warnings.warn("Couldn't retrieve source code for container of "
                              "type " + container_type.__name__ + ". It won't be checked "
                              "for correctness upon loading.")
                return
            if original_source != current_source:
                if container_type.dump_patches:
                    file_name = container_type.__name__ + '.patch'
                    diff = difflib.unified_diff(current_source.split('\n'),
                                                original_source.split('\n'),
                                                source_file,
                                                source_file, lineterm="")
                    lines = '\n'.join(diff)
                    try:
                        with open(file_name, 'a+') as f:
                            file_size = f.seek(0, 2)
                            f.seek(0)
                            if file_size == 0:
                                f.write(lines)
                            elif file_size != len(lines) or f.read() != lines:
                                raise IOError
                        msg = ("Saved a reverse patch to " + file_name + ". "
                               "Run `patch -p0 < " + file_name + "` to revert your "
                               "changes.")
                    except IOError:
                        msg = ("Tried to save a patch, but couldn't create a "
                               "writable file " + file_name + ". Make sure it "
                               "doesn't exist and your working directory is "
                               "writable.")
                else:
                    msg = ("you can retrieve the original source code by "
                           "accessing the object's source attribute or set "
                           "`torch.nn.Module.dump_patches = True` and use the "
                           "patch tool to revert the changes.")
                msg = ("source code of class '{}' has changed. {}"
                       .format(torch.typename(container_type), msg))
                warnings.warn(msg, SourceChangeWarning)
    
        def legacy_load(f):
            deserialized_objects = {}
    
            def persistent_load(saved_id):
                if isinstance(saved_id, tuple):
                    # Ignore containers that don't have any sources saved
                    if all(saved_id[1:]):
                        _check_container_source(*saved_id)
                    return saved_id[0]
                return deserialized_objects[int(saved_id)]
    
            with closing(tarfile.open(fileobj=f, mode='r:', format=tarfile.PAX_FORMAT)) as tar, \
                    mkdtemp() as tmpdir:
    
                tar.extract('storages', path=tmpdir)
                with open(os.path.join(tmpdir, 'storages'), 'rb', 0) as f:
                    num_storages = pickle_module.load(f, **pickle_load_args)
                    for i in range(num_storages):
                        args = pickle_module.load(f, **pickle_load_args)
                        key, location, storage_type = args
                        obj = storage_type._new_with_file(f)
                        obj = restore_location(obj, location)
                        deserialized_objects[key] = obj
    
                    storage_views = pickle_module.load(f, **pickle_load_args)
                    for target_cdata, root_cdata, offset, size in storage_views:
                        root = deserialized_objects[root_cdata]
                        deserialized_objects[target_cdata] = root[offset:offset + size]
    
                tar.extract('tensors', path=tmpdir)
                with open(os.path.join(tmpdir, 'tensors'), 'rb', 0) as f:
                    num_tensors = pickle_module.load(f, **pickle_load_args)
                    for _ in range(num_tensors):
                        args = pickle_module.load(f, **pickle_load_args)
                        key, storage_id, original_tensor_type = args
                        storage = deserialized_objects[storage_id]
                        tensor_type = storage_to_tensor_type(storage)
                        ndim, = struct.unpack('<i', f.read(4))
                        # skip next 4 bytes; legacy encoding treated ndim as 8 bytes
                        f.read(4)
                        size = struct.unpack('<{}q'.format(ndim), f.read(8 * ndim))
                        stride = struct.unpack('<{}q'.format(ndim), f.read(8 * ndim))
                        storage_offset, = struct.unpack('<q', f.read(8))
                        tensor = tensor_type().set_(storage, storage_offset, size, stride)
                        deserialized_objects[key] = tensor
    
                pickle_file = tar.extractfile('pickle')
                unpickler = pickle_module.Unpickler(pickle_file, **pickle_load_args)
                unpickler.persistent_load = persistent_load
                result = unpickler.load()
                return result
    
        deserialized_objects = {}
    
        def persistent_load(saved_id):
            assert isinstance(saved_id, tuple)
            typename = _maybe_decode_ascii(saved_id[0])
            data = saved_id[1:]
    
            if typename == 'module':
                # Ignore containers that don't have any sources saved
                if all(data[1:]):
                    _check_container_source(*data)
                return data[0]
            elif typename == 'storage':
                data_type, root_key, location, size, view_metadata = data
                location = _maybe_decode_ascii(location)
                if root_key not in deserialized_objects:
                    obj = data_type(size)
                    obj._torch_load_uninitialized = True
                    deserialized_objects[root_key] = restore_location(obj, location)
                storage = deserialized_objects[root_key]
                if view_metadata is not None:
                    view_key, offset, view_size = view_metadata
                    if view_key not in deserialized_objects:
                        deserialized_objects[view_key] = storage[offset:offset + view_size]
                    return deserialized_objects[view_key]
                else:
                    return storage
            else:
                raise RuntimeError("Unknown saved id type: %s" % saved_id[0])
    
        _check_seekable(f)
        f_should_read_directly = _should_read_directly(f)
    
        if f_should_read_directly and f.tell() == 0:
            # legacy_load requires that f has fileno()
            # only if offset is zero we can attempt the legacy tar file loader
            try:
                return legacy_load(f)
            except tarfile.TarError:
                if _is_zipfile(f):
                    # .zip is used for torch.jit.save and will throw an un-pickling error here
                    raise RuntimeError("{} is a zip archive (did you mean to use torch.jit.load()?)".format(f.name))
                # if not a tarfile, reset file offset and proceed
                f.seek(0)
    
>       magic_number = pickle_module.load(f, **pickle_load_args)
E       TypeError: file must have 'read', 'readinto' and 'readline' attributes

../../miniconda3/envs/skorch-2/lib/python3.8/site-packages/torch/serialization.py:692: TypeError
________________________ test_gridsearch_classification ________________________

    def test_gridsearch_classification():
        estimator = create_model()
        disparity_moment = DemographicParity()
    
>       ptc.run_gridsearch_classification(estimator, disparity_moment)

test_pytorch.py:93: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
package_test_common.py:80: in run_gridsearch_classification
    unmitigated = copy.deepcopy(estimator)
../../miniconda3/envs/skorch-2/lib/python3.8/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
../../miniconda3/envs/skorch-2/lib/python3.8/copy.py:271: in _reconstruct
    y.__setstate__(state)
../../miniconda3/envs/skorch-2/lib/python3.8/site-packages/skorch/net.py:1696: in __setstate__
    cuda_attrs = torch.load(f, **load_kwargs)
../../miniconda3/envs/skorch-2/lib/python3.8/site-packages/torch/serialization.py:529: in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

f = <tempfile.SpooledTemporaryFile object at 0x7fb628f31a90>
map_location = device(type='cpu')
pickle_module = <module 'pickle' from '/home/rge21/miniconda3/envs/skorch-2/lib/python3.8/pickle.py'>
pickle_load_args = {'encoding': 'utf-8'}
legacy_load = <function _legacy_load.<locals>.legacy_load at 0x7fb628fa8430>
persistent_load = <function _legacy_load.<locals>.persistent_load at 0x7fb628fa8550>
f_should_read_directly = True

    def _legacy_load(f, map_location, pickle_module, **pickle_load_args):
        deserialized_objects = {}
    
        restore_location = _get_restore_location(map_location)
    
        def _check_container_source(container_type, source_file, original_source):
            try:
                current_source = ''.join(get_source_lines_and_file(container_type)[0])
            except Exception:  # saving the source is optional, so we can ignore any errors
                warnings.warn("Couldn't retrieve source code for container of "
                              "type " + container_type.__name__ + ". It won't be checked "
                              "for correctness upon loading.")
                return
            if original_source != current_source:
                if container_type.dump_patches:
                    file_name = container_type.__name__ + '.patch'
                    diff = difflib.unified_diff(current_source.split('\n'),
                                                original_source.split('\n'),
                                                source_file,
                                                source_file, lineterm="")
                    lines = '\n'.join(diff)
                    try:
                        with open(file_name, 'a+') as f:
                            file_size = f.seek(0, 2)
                            f.seek(0)
                            if file_size == 0:
                                f.write(lines)
                            elif file_size != len(lines) or f.read() != lines:
                                raise IOError
                        msg = ("Saved a reverse patch to " + file_name + ". "
                               "Run `patch -p0 < " + file_name + "` to revert your "
                               "changes.")
                    except IOError:
                        msg = ("Tried to save a patch, but couldn't create a "
                               "writable file " + file_name + ". Make sure it "
                               "doesn't exist and your working directory is "
                               "writable.")
                else:
                    msg = ("you can retrieve the original source code by "
                           "accessing the object's source attribute or set "
                           "`torch.nn.Module.dump_patches = True` and use the "
                           "patch tool to revert the changes.")
                msg = ("source code of class '{}' has changed. {}"
                       .format(torch.typename(container_type), msg))
                warnings.warn(msg, SourceChangeWarning)
    
        def legacy_load(f):
            deserialized_objects = {}
    
            def persistent_load(saved_id):
                if isinstance(saved_id, tuple):
                    # Ignore containers that don't have any sources saved
                    if all(saved_id[1:]):
                        _check_container_source(*saved_id)
                    return saved_id[0]
                return deserialized_objects[int(saved_id)]
    
            with closing(tarfile.open(fileobj=f, mode='r:', format=tarfile.PAX_FORMAT)) as tar, \
                    mkdtemp() as tmpdir:
    
                tar.extract('storages', path=tmpdir)
                with open(os.path.join(tmpdir, 'storages'), 'rb', 0) as f:
                    num_storages = pickle_module.load(f, **pickle_load_args)
                    for i in range(num_storages):
                        args = pickle_module.load(f, **pickle_load_args)
                        key, location, storage_type = args
                        obj = storage_type._new_with_file(f)
                        obj = restore_location(obj, location)
                        deserialized_objects[key] = obj
    
                    storage_views = pickle_module.load(f, **pickle_load_args)
                    for target_cdata, root_cdata, offset, size in storage_views:
                        root = deserialized_objects[root_cdata]
                        deserialized_objects[target_cdata] = root[offset:offset + size]
    
                tar.extract('tensors', path=tmpdir)
                with open(os.path.join(tmpdir, 'tensors'), 'rb', 0) as f:
                    num_tensors = pickle_module.load(f, **pickle_load_args)
                    for _ in range(num_tensors):
                        args = pickle_module.load(f, **pickle_load_args)
                        key, storage_id, original_tensor_type = args
                        storage = deserialized_objects[storage_id]
                        tensor_type = storage_to_tensor_type(storage)
                        ndim, = struct.unpack('<i', f.read(4))
                        # skip next 4 bytes; legacy encoding treated ndim as 8 bytes
                        f.read(4)
                        size = struct.unpack('<{}q'.format(ndim), f.read(8 * ndim))
                        stride = struct.unpack('<{}q'.format(ndim), f.read(8 * ndim))
                        storage_offset, = struct.unpack('<q', f.read(8))
                        tensor = tensor_type().set_(storage, storage_offset, size, stride)
                        deserialized_objects[key] = tensor
    
                pickle_file = tar.extractfile('pickle')
                unpickler = pickle_module.Unpickler(pickle_file, **pickle_load_args)
                unpickler.persistent_load = persistent_load
                result = unpickler.load()
                return result
    
        deserialized_objects = {}
    
        def persistent_load(saved_id):
            assert isinstance(saved_id, tuple)
            typename = _maybe_decode_ascii(saved_id[0])
            data = saved_id[1:]
    
            if typename == 'module':
                # Ignore containers that don't have any sources saved
                if all(data[1:]):
                    _check_container_source(*data)
                return data[0]
            elif typename == 'storage':
                data_type, root_key, location, size, view_metadata = data
                location = _maybe_decode_ascii(location)
                if root_key not in deserialized_objects:
                    obj = data_type(size)
                    obj._torch_load_uninitialized = True
                    deserialized_objects[root_key] = restore_location(obj, location)
                storage = deserialized_objects[root_key]
                if view_metadata is not None:
                    view_key, offset, view_size = view_metadata
                    if view_key not in deserialized_objects:
                        deserialized_objects[view_key] = storage[offset:offset + view_size]
                    return deserialized_objects[view_key]
                else:
                    return storage
            else:
                raise RuntimeError("Unknown saved id type: %s" % saved_id[0])
    
        _check_seekable(f)
        f_should_read_directly = _should_read_directly(f)
    
        if f_should_read_directly and f.tell() == 0:
            # legacy_load requires that f has fileno()
            # only if offset is zero we can attempt the legacy tar file loader
            try:
                return legacy_load(f)
            except tarfile.TarError:
                if _is_zipfile(f):
                    # .zip is used for torch.jit.save and will throw an un-pickling error here
                    raise RuntimeError("{} is a zip archive (did you mean to use torch.jit.load()?)".format(f.name))
                # if not a tarfile, reset file offset and proceed
                f.seek(0)
    
>       magic_number = pickle_module.load(f, **pickle_load_args)
E       TypeError: file must have 'read', 'readinto' and 'readline' attributes

../../miniconda3/envs/skorch-2/lib/python3.8/site-packages/torch/serialization.py:692: TypeError
____________________ test_thresholdoptimizer_classification ____________________

    def test_thresholdoptimizer_classification():
        estimator = create_model()
    
>       ptc.run_thresholdoptimizer_classification(estimator)

test_pytorch.py:99: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
package_test_common.py:104: in run_thresholdoptimizer_classification
    unmitigated = copy.deepcopy(estimator)
../../miniconda3/envs/skorch-2/lib/python3.8/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
../../miniconda3/envs/skorch-2/lib/python3.8/copy.py:271: in _reconstruct
    y.__setstate__(state)
../../miniconda3/envs/skorch-2/lib/python3.8/site-packages/skorch/net.py:1696: in __setstate__
    cuda_attrs = torch.load(f, **load_kwargs)
../../miniconda3/envs/skorch-2/lib/python3.8/site-packages/torch/serialization.py:529: in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

f = <tempfile.SpooledTemporaryFile object at 0x7fb628f76610>
map_location = device(type='cpu')
pickle_module = <module 'pickle' from '/home/rge21/miniconda3/envs/skorch-2/lib/python3.8/pickle.py'>
pickle_load_args = {'encoding': 'utf-8'}
legacy_load = <function _legacy_load.<locals>.legacy_load at 0x7fb62905ed30>
persistent_load = <function _legacy_load.<locals>.persistent_load at 0x7fb62624b8b0>
f_should_read_directly = True

    def _legacy_load(f, map_location, pickle_module, **pickle_load_args):
        deserialized_objects = {}
    
        restore_location = _get_restore_location(map_location)
    
        def _check_container_source(container_type, source_file, original_source):
            try:
                current_source = ''.join(get_source_lines_and_file(container_type)[0])
            except Exception:  # saving the source is optional, so we can ignore any errors
                warnings.warn("Couldn't retrieve source code for container of "
                              "type " + container_type.__name__ + ". It won't be checked "
                              "for correctness upon loading.")
                return
            if original_source != current_source:
                if container_type.dump_patches:
                    file_name = container_type.__name__ + '.patch'
                    diff = difflib.unified_diff(current_source.split('\n'),
                                                original_source.split('\n'),
                                                source_file,
                                                source_file, lineterm="")
                    lines = '\n'.join(diff)
                    try:
                        with open(file_name, 'a+') as f:
                            file_size = f.seek(0, 2)
                            f.seek(0)
                            if file_size == 0:
                                f.write(lines)
                            elif file_size != len(lines) or f.read() != lines:
                                raise IOError
                        msg = ("Saved a reverse patch to " + file_name + ". "
                               "Run `patch -p0 < " + file_name + "` to revert your "
                               "changes.")
                    except IOError:
                        msg = ("Tried to save a patch, but couldn't create a "
                               "writable file " + file_name + ". Make sure it "
                               "doesn't exist and your working directory is "
                               "writable.")
                else:
                    msg = ("you can retrieve the original source code by "
                           "accessing the object's source attribute or set "
                           "`torch.nn.Module.dump_patches = True` and use the "
                           "patch tool to revert the changes.")
                msg = ("source code of class '{}' has changed. {}"
                       .format(torch.typename(container_type), msg))
                warnings.warn(msg, SourceChangeWarning)
    
        def legacy_load(f):
            deserialized_objects = {}
    
            def persistent_load(saved_id):
                if isinstance(saved_id, tuple):
                    # Ignore containers that don't have any sources saved
                    if all(saved_id[1:]):
                        _check_container_source(*saved_id)
                    return saved_id[0]
                return deserialized_objects[int(saved_id)]
    
            with closing(tarfile.open(fileobj=f, mode='r:', format=tarfile.PAX_FORMAT)) as tar, \
                    mkdtemp() as tmpdir:
    
                tar.extract('storages', path=tmpdir)
                with open(os.path.join(tmpdir, 'storages'), 'rb', 0) as f:
                    num_storages = pickle_module.load(f, **pickle_load_args)
                    for i in range(num_storages):
                        args = pickle_module.load(f, **pickle_load_args)
                        key, location, storage_type = args
                        obj = storage_type._new_with_file(f)
                        obj = restore_location(obj, location)
                        deserialized_objects[key] = obj
    
                    storage_views = pickle_module.load(f, **pickle_load_args)
                    for target_cdata, root_cdata, offset, size in storage_views:
                        root = deserialized_objects[root_cdata]
                        deserialized_objects[target_cdata] = root[offset:offset + size]
    
                tar.extract('tensors', path=tmpdir)
                with open(os.path.join(tmpdir, 'tensors'), 'rb', 0) as f:
                    num_tensors = pickle_module.load(f, **pickle_load_args)
                    for _ in range(num_tensors):
                        args = pickle_module.load(f, **pickle_load_args)
                        key, storage_id, original_tensor_type = args
                        storage = deserialized_objects[storage_id]
                        tensor_type = storage_to_tensor_type(storage)
                        ndim, = struct.unpack('<i', f.read(4))
                        # skip next 4 bytes; legacy encoding treated ndim as 8 bytes
                        f.read(4)
                        size = struct.unpack('<{}q'.format(ndim), f.read(8 * ndim))
                        stride = struct.unpack('<{}q'.format(ndim), f.read(8 * ndim))
                        storage_offset, = struct.unpack('<q', f.read(8))
                        tensor = tensor_type().set_(storage, storage_offset, size, stride)
                        deserialized_objects[key] = tensor
    
                pickle_file = tar.extractfile('pickle')
                unpickler = pickle_module.Unpickler(pickle_file, **pickle_load_args)
                unpickler.persistent_load = persistent_load
                result = unpickler.load()
                return result
    
        deserialized_objects = {}
    
        def persistent_load(saved_id):
            assert isinstance(saved_id, tuple)
            typename = _maybe_decode_ascii(saved_id[0])
            data = saved_id[1:]
    
            if typename == 'module':
                # Ignore containers that don't have any sources saved
                if all(data[1:]):
                    _check_container_source(*data)
                return data[0]
            elif typename == 'storage':
                data_type, root_key, location, size, view_metadata = data
                location = _maybe_decode_ascii(location)
                if root_key not in deserialized_objects:
                    obj = data_type(size)
                    obj._torch_load_uninitialized = True
                    deserialized_objects[root_key] = restore_location(obj, location)
                storage = deserialized_objects[root_key]
                if view_metadata is not None:
                    view_key, offset, view_size = view_metadata
                    if view_key not in deserialized_objects:
                        deserialized_objects[view_key] = storage[offset:offset + view_size]
                    return deserialized_objects[view_key]
                else:
                    return storage
            else:
                raise RuntimeError("Unknown saved id type: %s" % saved_id[0])
    
        _check_seekable(f)
        f_should_read_directly = _should_read_directly(f)
    
        if f_should_read_directly and f.tell() == 0:
            # legacy_load requires that f has fileno()
            # only if offset is zero we can attempt the legacy tar file loader
            try:
                return legacy_load(f)
            except tarfile.TarError:
                if _is_zipfile(f):
                    # .zip is used for torch.jit.save and will throw an un-pickling error here
                    raise RuntimeError("{} is a zip archive (did you mean to use torch.jit.load()?)".format(f.name))
                # if not a tarfile, reset file offset and proceed
                f.seek(0)
    
>       magic_number = pickle_module.load(f, **pickle_load_args)
E       TypeError: file must have 'read', 'readinto' and 'readline' attributes

../../miniconda3/envs/skorch-2/lib/python3.8/site-packages/torch/serialization.py:692: TypeError
=========================== short test summary info ============================
FAILED test_pytorch.py::test_expgrad_classification - TypeError: file must ha...
FAILED test_pytorch.py::test_gridsearch_classification - TypeError: file must...
FAILED test_pytorch.py::test_thresholdoptimizer_classification - TypeError: f...
============================== 3 failed in 5.90s ===============================
```

Reply:
Then pip freeze:
```python
attrs @ file:///tmp/build/80754af9/attrs_1604765588209/work
certifi==2020.6.20
cffi==1.14.0
fairlearn @ file:///home/rge21/fairlearn
iniconfig @ file:///tmp/build/80754af9/iniconfig_1602780191262/work
joblib @ file:///tmp/build/80754af9/joblib_1601912903842/work
mkl-fft==1.2.0
mkl-random==1.1.1
mkl-service==2.3.0
more-itertools @ file:///tmp/build/80754af9/more-itertools_1605111547926/work
numpy @ file:///tmp/build/80754af9/numpy_and_numpy_base_1603570489231/work
packaging==20.4
pandas==1.1.4
pip==20.2.4
pluggy==0.13.1
py @ file:///tmp/build/80754af9/py_1593446248552/work
pycparser @ file:///tmp/build/80754af9/pycparser_1594388511720/work
pyparsing==2.4.7
pytest==0.0.0
python-dateutil==2.8.1
pytz==2020.4
scikit-learn @ file:///tmp/build/80754af9/scikit-learn_1598376899566/work
scipy @ file:///tmp/build/80754af9/scipy_1597686649129/work
setuptools==50.3.1.post20201107
six @ file:///tmp/build/80754af9/six_1605205327372/work
skorch @ file:///home/conda/feedstock_root/build_artifacts/skorch_1600488572036/work
tabulate==0.8.7
threadpoolctl @ file:///tmp/tmp9twdgx9k/threadpoolctl-2.1.0-py3-none-any.whl
toml @ file:///tmp/build/80754af9/toml_1592853716807/work
torch==1.4.0
tqdm @ file:///tmp/build/80754af9/tqdm_1605303662894/work
wheel==0.35.1
```

and conda environment:

```python
# packages in environment at /home/rge21/miniconda3/envs/skorch-2:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
_pytorch_select           0.2                       gpu_0  
attrs                     20.3.0             pyhd3eb1b0_0  
blas                      1.0                         mkl  
ca-certificates           2020.10.14                    0  
certifi                   2020.6.20          pyhd3eb1b0_3  
cffi                      1.14.0           py38h2e261b9_0  
cpuonly                   1.0                           0    pytorch
cudatoolkit               10.1.243             h6bb024c_0  
cudnn                     7.6.5                cuda10.1_0  
fairlearn                 0.5.1.dev0               pypi_0    pypi
iniconfig                 1.1.1                      py_0  
intel-openmp              2020.2                      254  
joblib                    0.17.0                     py_0  
libedit                   3.1.20191231         h14c3975_1  
libffi                    3.2.1             hf484d3e_1007  
libgcc-ng                 9.1.0                hdf63c60_0  
libgfortran-ng            7.3.0                hdf63c60_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
mkl                       2020.2                      256  
mkl-service               2.3.0            py38he904b0f_0  
mkl_fft                   1.2.0            py38h23d657b_0  
mkl_random                1.1.1            py38h0573a6f_0  
more-itertools            8.6.0              pyhd3eb1b0_0  
ncurses                   6.2                  he6710b0_1  
ninja                     1.10.1           py38hfd86e86_0  
numpy                     1.19.2           py38h54aff64_0  
numpy-base                1.19.2           py38hfa32c7d_0  
openssl                   1.1.1h               h7b6447c_0  
packaging                 20.4                       py_0  
pandas                    1.1.4                    pypi_0    pypi
pip                       20.2.4           py38h06a4308_0  
pluggy                    0.13.1                   py38_0  
py                        1.9.0                      py_0  
pycparser                 2.20                       py_2  
pyparsing                 2.4.7                      py_0  
pytest                    6.1.1                    py38_0  
python                    3.8.0                h0371630_2  
python-dateutil           2.8.1                    pypi_0    pypi
pytorch                   1.4.0           cuda101py38h02f0884_0  
pytz                      2020.4                   pypi_0    pypi
readline                  7.0                  h7b6447c_5  
scikit-learn              0.23.2           py38h0573a6f_0  
scipy                     1.5.2            py38h0b6359f_0  
setuptools                50.3.1           py38h06a4308_1  
six                       1.15.0           py38h06a4308_0  
skorch                    0.9.0              pyh7b7c402_0    conda-forge
sqlite                    3.33.0               h62c20be_0  
tabulate                  0.8.7                    py38_0  
threadpoolctl             2.1.0              pyh5ca1d4c_0  
tk                        8.6.10               hbc83047_0  
toml                      0.10.1                     py_0  
tqdm                      4.51.0             pyhd3eb1b0_0  
wheel                     0.35.1             pyhd3eb1b0_0  
xz                        5.2.5                h7b6447c_0  
zlib                      1.2.11               h7b6447c_3  
```

Reply:
This is on an Ubuntu 18.04 LTS VM.

Reply:
Oh I am so sorry, some changes were not commited. The issue was python 3.8, Pytorch only works for 3.7.

Reply:
Got your latest changes, and this works now.